### DISCUSSION TITLE
Question about MessagePassing Class

### DISCUSSION THREAD

**Discussion by hkim716** [pyg-team/pytorch_geometric#2957] [Q&A] [answered]
Hi, Matt. 
I'm trying to understand `MessagePassing` Class used in PyG. 
[This](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.message_passing.MessagePassing) page helped me a lot, but I have some questions.
When I see each code of convolution layers in PyG, they define `forward` function, and in the `forward` function, there is `propagate()` function, which internally calls the `message()`, `aggregate()` and `update()` functions.
I found most `conv` layers in PyG have `message()` functions there, but I cannot easily find `aggregate()` and `update()` functions. For example, in `spline_conv.py`, there is `forward` function, which calls `propagate()`. And `propagate()` will call `message()`. 

Here are my questions.

1. Are `aggregate()` and `update()` hidden by default?
2. I found some `conv` layers have `message_and_aggregate()` function. Does this function take the output from `message()` by default?
3. In my understanding, the entire `MessagePassing` operation has some orders like this : 
`forward()`->`propagate()`->`message()`->`( message_and_aggregate() )`->`aggregate()`->`update()`
Am I right?

**Comment by rusty1s** [comment, marked as answer]
1. [`aggregate()`](https://github.com/rusty1s/pytorch_geometric/blob/master/torch_geometric/nn/conv/message_passing.py#L273-L291) and [`update()`](https://github.com/rusty1s/pytorch_geometric/blob/master/torch_geometric/nn/conv/message_passing.py#L303-L310) have default implementations. For example, by default, `aggregate` will aggregate incoming messages according to the `aggr` attribute (`"add"`, `"mean"` or `"max"`) of the `MessagePassing` class. `update()` just returns the input by default.
2. `message_and_aggregate()` fuses `message` and `aggregate` computation into a single step, which is useful in order to allow for more memory-efficient GNN operators. [This post](https://pytorch-geometric.readthedocs.io/en/latest/notes/sparse_tensor.html) tries to explain it in more detail.
3. That is mostly correct. If `message_and_aggregate` is not implemented, the order of operations is:
```
forward -> propagate -> message -> aggregate -> update
```
If `message_and_aggregate` is implemented, the order of operations is:
```
forward -> propagate -> message_and_aggregate -> update
```
